src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java

```diff
@@ -19,63 +27,186 @@
+
+        //bbspullrequests is ordered from newest to oldest
+        bbsPullRequests.map(pr -> {
+            if (pr.getUpdatedDate() > oldestUpdate.get()) {
+                updatePullRequest(serverId, pr);
+            }
+            return pr.getUpdatedDate();
+        }).filter(updated -> updated < oldestUpdate.get())
+                .findFirst();
+    }
+
+    @Override
+    public void removeClosedPullRequests(long date) {
+       pullRequests.values().forEach(repositoryStore -> repositoryStore.removeClosedPullRequests(date));
+    }
+
+    @Override
+    public boolean hasPRForRepository(String projectKey, String slug, String serverId) {
+        PullRequestStoreImpl.CacheKey key =
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        return pullRequests.getOrDefault(key, new RepositoryStore(new ConcurrentHashMap<>())).hasPR();
+    }
+
+    public static MinimalPullRequest convert(BitbucketPullRequest bbsPR) {
+        return new MinimalPullRequest(bbsPR.getId(), bbsPR.getState(), bbsPR.getFromRef().getDisplayId(),
+                bbsPR.getToRef().getDisplayId(), bbsPR.getUpdatedDate());
+    }
+
+    private static class RepositoryStore {
+
+        private static final MinimalPullRequest CLOSED_PR = new MinimalPullRequest(0, BitbucketPullState.DECLINED,
+                "", "", 0);
+        private ConcurrentMap<String, ConcurrentMap<String, MinimalPullRequest>> pullRequests;
+
+        public RepositoryStore(
+                ConcurrentMap<String, ConcurrentMap<String, MinimalPullRequest>> pullRequests) {
+            this.pullRequests = pullRequests;
+        }
+
+        public void closePullRequest(String fromBranch, String toBranch) {
+            pullRequests.getOrDefault(fromBranch, new ConcurrentHashMap<>()).getOrDefault(toBranch, CLOSED_PR)
+                    .setState(BitbucketPullState.DELETED);
+        }
+
+        public AtomicLong findOldestUpdate() {
+            AtomicLong oldestUpdate = new AtomicLong(0);
+            pullRequests.values().stream().forEach(fromBranch -> {
+                fromBranch.values().forEach(pr -> {
+                    if (oldestUpdate.get() == 0) {
+                        oldestUpdate.set(pr.getUpdatedDate());
+                    }
+                    if (pr.getUpdatedDate() < oldestUpdate.get()) {
+                        oldestUpdate.set(pr.getUpdatedDate());
+                    }
+                });
+            });
+            return oldestUpdate;
+        }
+
+        public MinimalPullRequest getPullRequest(String fromBranch, String toBranch) {
+            return pullRequests.getOrDefault(fromBranch, new ConcurrentHashMap<>()).get(toBranch);
+        }
+
+        public boolean hasOpenPullRequests(String branchName) {
+            return pullRequests.getOrDefault(branchName, new ConcurrentHashMap<>()).values().stream().filter(
```

Comment:
```suggestion
            return pullRequests.getOrDefault(branchName, new ConcurrentHashMap<>()).values().stream().anyMatch(
                    pr -> pr.getState() == BitbucketPullState.OPEN);
```
Does away with the need for the `findFirst()`